File: java/17.17.0_Multi_Search_LCCI.java
/*
Trie

执行用时：46 ms, 在所有 Java 提交中击败了85.24% 的用户
内存消耗：61.7 MB, 在所有 Java 提交中击败了28.18% 的用户
通过测试用例：32 / 32
*/

class Solution {
    class TrieNode {
        String end;
        TrieNode[] next = new TrieNode[26];
    }

    class Trie {
        TrieNode root;
        public Trie(String[] words) {
            root = new TrieNode();
            for (String word: words) {
                TrieNode node = root;
                for (char ch: word.toCharArray()) {
                    int i = ch - 'a';
                    if (node.next[i] == null) {
                        node.next[i] = new TrieNode();
                    }
                    node = node.next[i];
                }
                node.end = word;
            }
        }

        public List<String> search(String str) {
            TrieNode node = root;
            List<String> res = new ArrayList<>();
            for (char c: str.toCharArray()) {
                int i = c - 'a';
                if (node.next[i] == null) {
                    break;
                }
                node = node.next[i];
                if (node.end != null) {
                    res.add(node.end);
                }
            }
            return res;
        }
    }

    public int[][] multiSearch(String big, String[] smalls) {
        Trie trie = new Trie(smalls);
        Map<String, List<Integer>> hit = new HashMap<>();
        for (int i = 0; i < big.length(); i++) {
            List<String> matches = trie.search(big.substring(i));
            for (String word: matches) {
                if (!hit.containsKey(word)) {
                    hit.put(word, new ArrayList<>());
                }
                hit.get(word).add(i);
            }
        }

        int[][] res = new int[smalls.length][];
        for (int i = 0; i < smalls.length; i++) {
            List<Integer> list = hit.get(smalls[i]);
            if (list == null) {
                res[i] = new int[0];
                continue;
            }
            int size = list.size();
            res[i] = new int[size];
            for (int j = 0; j < size; ++j) {
                res[i][j] = list.get(j);
            }
        }
        return res;
    }
}

File: leetcode-cn/CrackingTheCodingInterview/17.17.0_Multi_Search_LCCI.py
'''
Brute Force
T: O(10^8)

执行用时：360 ms, 在所有 Python3 提交中击败了42.61% 的用户
内存消耗：34.4 MB, 在所有 Python3 提交中击败了28.41% 的用户
通过测试用例：32 / 32
'''
class Solution:
    def multiSearch(self, big: str, smalls: List[str]) -> List[List[int]]:

        def f(s):
            if not s:
                return []
            cur = []
            idx = big.find(s)
            while idx >= 0:
                # yield idx
                cur.append(idx)
                idx = big.find(s, idx + 1)
            return cur

        return [f(s) for s in smalls]


'''
Trie

执行用时：232 ms, 在所有 Python3 提交中击败了68.75% 的用户
内存消耗：32.8 MB, 在所有 Python3 提交中击败了67.61% 的用户
通过测试用例：32 / 32
'''
class Trie:
    def __init__(self, words):
        self.d = {}
        for word in words:
            t = self.d
            for ch in word:
                if ch not in t:
                    t[ch] = {}
                t = t[ch]
            # 因为Trie里每个孩子都是一个小写字母,所以这里用'end'不影响
            t['end'] = word
    
    def search(self, str):
        t = self.d
        res = []
        for ch in str:
            if ch not in t:
                break
            t = t[ch]
            if 'end' in t:
                res.append(t['end'])
        return res

class Solution:
    def multiSearch(self, big: str, smalls: List[str]) -> List[List[int]]:
        trie = Trie(smalls)
        hit = defaultdict(list)
        for i in range(len(big)):
            matches = trie.search(big[i: ])
            for m in matches:
                hit[m].append(i)

        return [hit[s] for s in smalls]

